crates/iota-data-ingestion-core/src/reducer.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Mysten Labs, Inc.
+// Modifications Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+
+use async_trait::async_trait;
+use futures::StreamExt;
+use iota_types::messages_checkpoint::CheckpointSequenceNumber;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+
+use crate::{
+    IngestionError, IngestionResult, MAX_CHECKPOINTS_IN_PROGRESS, Worker,
+    worker_pool::WorkerPoolStatus,
+};
+
+#[async_trait]
+pub trait Reducer<Mapper: Worker>: Send + Sync {
+    async fn commit(&self, batch: Vec<Mapper::Message>) -> Result<(), Mapper::Error>;
+
+    fn should_close_batch(
+        &self,
+        _batch: &[Mapper::Message],
+        next_item: Option<&Mapper::Message>,
+    ) -> bool {
+        next_item.is_none()
+    }
+}
+
+/// Processes worker messages and manages batching through a reducer.
+///
+/// This function is the core of the reduction pipeline, handling message
+/// batching, watermark tracking, and progress reporting. It maintains message
+/// order by checkpoint sequence number and applies batching logic through the
+/// provided reducer.
+///
+/// # Message Processing Flow
+///
+/// 1. Receives messages in chunks up to [`MAX_CHECKPOINTS_IN_PROGRESS`]
+/// 2. Maintains message order using checkpoint sequence numbers
+/// 3. Batches messages according to reducer's [`Reducer::should_close_batch`]
+///    policy and after that its committed
+/// 4. Progress is updated after each commit
+/// 5. Reports progress back to the executor
+///
+/// # Shutdown Behavior
+///
+/// The function will gracefully exit when receiving a shutdown signal,
+/// ensuring no data loss for processed messages.
+pub(crate) async fn reduce<W: Worker>(
+    task_name: String,
+    mut current_checkpoint_number: CheckpointSequenceNumber,
+    watermark_receiver: mpsc::Receiver<(CheckpointSequenceNumber, W::Message)>,
+    executor_progress_sender: mpsc::Sender<WorkerPoolStatus>,
+    reducer: Box<dyn Reducer<W>>,
+) -> IngestionResult<()> {
+    // convert to a stream of MAX_CHECKPOINTS_IN_PROGRESS size. This way, each
+    // iteration of the loop will process all ready messages
+    let mut stream =
+        ReceiverStream::new(watermark_receiver).ready_chunks(MAX_CHECKPOINTS_IN_PROGRESS);
+    // store unprocessed progress messages from workers.
+    let mut unprocessed = HashMap::new();
+    // buffer to accumulate results before passing them to the reducer.
+    // The size of this batch is dynamically determined by the reducer's
+    // `should_close_batch` method.
+    let mut batch = vec![];
+    // track the latest processed checkpoint number for reporting progress
```

Comment:
Given the analysis, this should be "track the next unprocessed checkpoint..." or sth like this.

Reply:
fixed: https://github.com/iotaledger/iota/pull/5547/commits/5513315a63d0c15ae6c3142d0290d2a862d038e2